Simplenote/src/main/java/com/automattic/simplenote/utils/AccountVerificationWatcher.java

```diff
@@ -18,57 +18,63 @@
  * have (and probably was) updated from another device.
  */
 public class AccountVerificationWatcher implements Bucket.OnNetworkChangeListener<Account> {
-    private enum Status {
-        UNKNOWN,
+    public enum Status {
+        SENT_EMAIL,
         UNVERIFIED,
         VERIFIED,
     }
 
+    public interface VerificationStateListener {
+        void onUpdate(Status status);
+    }
+
     private final Simplenote simplenote;
-    private Status currentStatus = Status.UNKNOWN;
+    private final List<VerificationStateListener> listeners = new ArrayList<>();
 
     public AccountVerificationWatcher(Simplenote simplenote) {
         this.simplenote = simplenote;
     }
 
-    public void clearStatus() {
-        currentStatus = Status.UNKNOWN;
+    public void addVerificationStateListener(VerificationStateListener listener) {
+        listeners.add(listener);
     }
 
-    @Override
-    public void onNetworkChange(Bucket<Account> bucket, Bucket.ChangeType type, String key) {
-        // If we already know the answer, we avoid computing it again
-        if (currentStatus != Status.UNKNOWN) {
-            return;
+    private void notifyListeners(Status status) {
+        for (VerificationStateListener listener : listeners) {
+            listener.onUpdate(status);
         }
+    }
+
+    @Override
+    public void onNetworkChange(final Bucket<Account> bucket, Bucket.ChangeType type, String key) {
+        // If the account is removed, the status is changed to UNVERIFIED immediately since the account will not be
+        // available
+       if (type == Bucket.ChangeType.REMOVE) {
+            notifyListeners(Status.UNVERIFIED);
+           return;
+       }
 
         String email = simplenote.getUserEmail();
-        if (type != Bucket.ChangeType.INDEX || email == null) {
+        if ((type != Bucket.ChangeType.INDEX && type != Bucket.ChangeType.MODIFY) || email == null) {
```

Comment:
unfortunately we want to monitor for `INSERT` here too as well. we can safely ignore `RESET`. we might want to change these conditionals as a result.

```java
if (type == Bucket.ChangeType.RESET || email == null) {
    return;
}

// get the entity

// set verification based on presence and validity of entity
```

Reply:
I'm worried we'll have to track the status to avoid notifying on changes if we haven't changed status. if we have to avoid double-sending the notification then I guess we go back to what you had, but I think we can still focus on the `onChange` interface to keep that state-tracking in this module, to avoid exposing `UNKNOWN`

Reply:
> unfortunately we want to monitor for INSERT here too as well. we can safely ignore RESET. we might want to change these conditionals as a result.

Question(❓ ): Will the `key` will not be `null for `INSERT` as well as for `REMOVE` (question above)?

Reply:
> I'm worried we'll have to track the status to avoid notifying on changes if we haven't changed status. if we have to avoid double-sending the notification then I guess we go back to what you had, but I think we can still focus on the onChange interface to keep that state-tracking in this module, to avoid exposing UNKNOWN

We can keep the status (`currentState`) and just notify the listener when the status changes: `if (newState != currentState)` with `currentState = UNKNOWN` as default value.

Reply:
> Will the key will not be null for INSERT as well as for REMOVE (question above)?

Easy to test 😄 

But no they won't be `null` - they will have the `entity_id` as a string value. Wouldn't make much sense to always and only pass `null` as an argument I suppose…

Reply:
Yes, that is what I though later, put a listener on the notes bucket and observe the behavior on`INSERT` and `MODIFY`

Reply:
> Wouldn't make much sense to always and only pass null as an argument I suppose…

We could improve the interface and apply the `Interface Segregation Principle` since this interface may be doing more than one thing 😜

Reply:
> The interface is handling two types of events, one for the whole bucket and one for specific keys.

Yes, this is a weakness of the current design. It's actually because of issues like this why I put my attention into code reviews to encourage that we deeply consider the new interfaces we add so as to make a positive turn in the app's design.

Reply:
Double 👍  to make it even 🤣